File: problem_solving/linkedlist/py/2_add_two_numbers.py
'''
https://leetcode.com/problems/add-two-numbers/description/?envType=study-plan-v2&envId=top-interview-150

#meta
'''
from linkedlist import ListNode


def add_two_numbers(l1:ListNode,l2: ListNode):
    carry = 0
    l3 = ListNode(0)
    head = l3
    while l1 or l2 or carry:    
        add = 0
        if l1:
            add += l1.val
            l1 = l1.next
        if l2:
            add += l2.val
            l2 = l2.next
        add += carry
        carry = add // 10  #missed the // double
        val = add % 10
        l3.next = ListNode(val)
        l3 = l3.next
    return head.next


A1 = ListNode(2)
B1 = ListNode(4)
C1 = ListNode(3)
A1.next = B1
B1.next = C1
print("A1: " + A1.traverse())

A2 = ListNode(5)
B2 = ListNode(6)
C2 = ListNode(4)
A2.next = B2
B2.next = C2
print("A2: " + A2.traverse())
#l1 = [2,4,3]
#l2 = [5,6,4]
#Output: [7,0,8]
l3 = add_two_numbers(A1, A2)
print("l3: " + l3.traverse())

#------------------------------

l1 = [0]
l2 = [0]
#Output: [0]
l3 = add_two_numbers(A1, A2)
#print("l3: " + l3.traverse())
A1 = ListNode(0)
print("A1: " + A1.traverse())
A2 = ListNode(0)
l3 = add_two_numbers(A1, A2)
print("l3: " + l3.traverse())

#----------------------------------


#l1 = [9,9,9,9,9,9,9]
#l2 = [9,9,9,9]
#Output: [8,9,9,9,0,0,0,1]
A1 = ListNode(9)
B1 = ListNode(9)
C1 = ListNode(9)
D1 = ListNode(9)
E1 = ListNode(9)
F1 = ListNode(9)
G1 = ListNode(9)

A1.next = B1
B1.next = C1
C1.next = D1
D1.next = E1
E1.next = F1
F1.next = G1

print("A1: " + A1.traverse())

A2 = ListNode(9)
B2 = ListNode(9)
C2 = ListNode(9)
D2 = ListNode(9)
A2.next = B2
B2.next = C2
C2.next = D2
print("A2: " + A2.traverse())
l3 = add_two_numbers(A1, A2)
print("l3: " + l3.traverse())

File: problem_solving/linkedlist/cpp/2_add_two_numbers.cpp
/*
https://leetcode.com/problems/add-two-numbers/description/?envType=study-plan-v2&envId=top-interview-150

add n1, n2 and carry

*/

#include<string>
#include<sstream>
#include<vector>
#include<iostream>
#include "../../utils/utils.h"
#include "../../common_header.h"
using namespace std;


ListNode* add_two_numbers(ListNode* a, ListNode* b)  {
    ListNode* result = new ListNode(0);
    auto node = result;
    int carry = 0;
    while(a || b || carry>0) {
        int sum = carry;
        if(a) {
            sum += a->val;
        }
        if (b) {
            sum += b->val;
        }
        int num = sum % 10;
        carry = sum / 10;
        node->next = new ListNode(num);
        node = node->next;
        if(a) {
            a = a->next;
        }
        if(b) {
            b = b->next;
        }
    }   
    ListNode* remaining = a;    
    if (b) {
        remaining = b;
    }
    node->next = remaining;
    return result->next;
}



int main() {
    ListNode* a = new ListNode(2);
    a->next = new ListNode(4);
    a->next->next = new ListNode(3);
    print_linked_list(a);

    ListNode* b = new ListNode(5);
    b->next = new ListNode(6);
    b->next->next = new ListNode(4);
    print_linked_list(b);

    auto result = add_two_numbers(a, b);
    std::cout << "result: ";
    print_linked_list(result);

    //case II
    a = new ListNode(0);
    print_linked_list(a);

    b = new ListNode(0);
    print_linked_list(b);

    result = add_two_numbers(a, b);
    std::cout << "result: ";
    print_linked_list(result);

    //case III
    a = new ListNode(9);
    a->next = new ListNode(9);
    a->next->next = new ListNode(9);
    a->next->next->next = new ListNode(9);
    a->next->next->next->next = new ListNode(9);
    a->next->next->next->next->next = new ListNode(9);
    a->next->next->next->next->next->next = new ListNode(9);
    print_linked_list(a);

    b = new ListNode(9);
    b->next = new ListNode(9);
    b->next->next = new ListNode(9);
    b->next->next->next = new ListNode(9);
    print_linked_list(b);

    result = add_two_numbers(a, b);
    std::cout << "result: ";
    print_linked_list(result);
}

